File: Data Structures & Algorithms/combination-target-sum/submission-2.cpp
class Solution {
public:
void solve(int index,vector<int>& curr,vector<vector<int>>& res,vector<int>& nums,int target){
    if(index==nums.size()){
        if(target==0)
        res.push_back(curr);
        return;
    }
    if(nums[index]<=target){
        curr.push_back(nums[index]);
        solve(index,curr,res,nums,target-nums[index]);
        curr.pop_back();
    }
    solve(index+1,curr,res,nums,target);
}
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        vector<int> curr;
        solve(0,curr,res,nums,target);
        return res;
    }
};

File: Data Structures & Algorithms/target-sum/submission-2.java
class Solution {
    int solve(int index,int sum,int target,int[] nums,int[][] dp){
        if(index==nums.length){
            if(sum==target) return 1;
            return 0;
        }
        if(dp[index][sum+1000]!=-1) return dp[index][sum+1000];
        int add=solve(index+1,sum+nums[index],target,nums,dp);
        int subtract=solve(index+1,sum-nums[index],target,nums,dp);
        return dp[index][sum+1000]=add+subtract;
    }
    public int findTargetSumWays(int[] nums, int target) {
        int[][] dp=new int[nums.length][2001];
        for(int i=0;i<nums.length;i++) Arrays.fill(dp[i],-1);
        return solve(0,0,target,nums,dp);
    }
}
